## Strings
Strings are generally non numericall datatypes that are used to store variables like "Names" , "Address" "Symbols" etc


For Eg:

`a = "Hello"`

### How does "+" operator differs from numerical to string operation 


For Eg :

```
a = 10
b = "10"
```

What will be the result below for all three scenarios 

scenario - 1 : Number + Number 

```
c = 5
print(a+c) 
```

scenario - 2 : String + String

```
c = "5"
print(b+c) 
```

Scenaio - 3 : NUmber + string 
```
f = 5
print("b" + f)
```

In [ ]:
### Scenario 1
a = 10
b = "10"
c = 5
print(a+c) 


15


In [ ]:
## scenario 2 
c = "5"
print(b+c) 

105


In [ ]:
## Scenario3 
f = 5
print("b" + f)

TypeError: ignored

## Note about String addition

When two strings are added, string `concatenation` takes place. String concatenation is joining of two strings with each other. 
For Eg:
`"Hello" + "World"` will yield "HelloWorld"


Note: You can concatenate ("+") only a string with a string , not a string with a number 

## Excerise - 1 
--- 

### Print Hello user 

The objective is to write a function named hello(), which accepts the name of the user as input argument and it returns the string as "Hello <username>"

for eg :     
```
k = hello("Magnus") 
print(k)
```
should print output as 
```
Hello Magnus
```


In [ ]:
#@title Solution - Excersise 1
def Hello(name):
  # a = "Hello "
  # b= name
  ## return a +b

  return "Hello " + name

## Exercise 2
---

### Print user name with their age 

Define a function `HelloAge`
- Input Arguments to the functions 
  - name of the user 
  - age of the user 
- Return 
   The method should return the final string
- Output 
For Eg : `HelloAge("Magnus",20)` should provide output as `Hi Magnus, your age is 20`

In [ ]:
#@title Solution - Problem 2
def HelloAge(name,age):
  return "Hello " + name + ", your age is " + str(age)

HelloAge("Magnus", 5)

'Hello Magnus, your age is 5'

### Concatenation of String with numbers 
---


If we have to concatenate a non string type ( such as numbers to a string ) then we should typecast ( convert ) the variable into a string and then append it. 


### Type casting ( Converting datatype )

Typecasting is just changing the datatype of the given variable to another data type 

For Eg: If we need to change a number to string 
```
a = 10
```

now 

```
a= str(a)
```

will change the value of a to a string . i.e 
```
a = "10"
```

--- 
---

Similarly we can typecast  a string to number, given it is a proper numerical value. 

```
k = "1980"
```

now we can typecast it back to number by using 
```
j = int(k)
```

now, j will be a numerical datatype , So we can check that by adding another number to it
```
>>> print(j+4)
    1984
```



#### Note :
Only strings with proper numerical values can be typecasted into a int or float. 

For eg : ` b = int("1875") ` is a valid conversion. However , `k = int("sdfsdfsd")` is not a valid conversion and it will throw error

In [ ]:
## String to int
a = "1776"
b = int(a) + 100
print(b)

1876


In [ ]:
### int to string
a = 1900
b = str(a)
c = "Hello "
print( c  + b)



Hello 1900


In [ ]:
### String to int, invalid conversion 
a = "sdfsdfsdF"
b = int(a)
print(b)

ValueError: ignored

In [ ]:
## String to a float 
a = "10.20"
b = float(a)
print(b+1.30)

11.5


With the above knowlegde lets correct the function below to produce proper output 



In [ ]:
def HelloAge(name,age):
  return "Hello " + name + ", your age is " + age

HelloAge("Magnus", 5)

I have two lists ,one with all piece names and another with all points 

piece = ["Rook","Pawn","bishop"]
points = [5,1,3]


write a function, which accepts these two lists as input arguments and then **print** the following 

```
Rook is worth 5 Points 
Pawn is worth 1 Points
Bishop is worth 3 points
```


## Hint

Since we are going to work on two lists at once.
So instead of using a forloop directly as `for p in piece` , use index based loop such as `for i in range(3):` and then use the indexes to access the value in every loop . Like `piece[i]`

In [ ]:
#idk2
piece = ["Rook","Pawn","Bishop"]
points = [5,1,3]
def Chess(piece,points):
   for i in range(len(points)):
      a = piece[i]
      b = points[i]
      print(str(a) + " is worth " + str(b) + " Points")

In [ ]:
Chess(piece,points)

Rook is worth 5 Points
Pawn is worth 1 Points
Bishop is worth 3 Points


### What is the Problem in the above method ?

1, We need to break out print statemens either by using a `,` or an concatenation `+` to substitute variables within our function 
2, Further, we need to typecast ( convert into another datatype) integer ( non string datatypes) into string before performing concatenation 

#### Variable Substitution 

Eg: 

```
age=10
## if i have to ptint your age is 10
## then I need to substitute the value of the variable "age" into my print statement 
print("Your Age is " , str(age))
print("Your Age is " + str(age))
```


What if we can write a single string , which can perform both of the following 
- Variable Substitution 
- Type Conversion 


So Thats where , Python FStrings are usefull 
---


# F-Strings Python 


Fstrings in python are very similar to the normal strings , However we will just add an `f` before the delcaration of the string 


For Eg:
```
##Rather than declaring a simple string like 
### Name =  'Magnus'
Name = f'Magnus'      ## Example of fstring
```

### How is this usefull ?

Fstrings has a feature to substitute variables inside a string using a `{}`
. So lets say if we have to substitute a variable called `name` in our string, then we can use 

#### Syntax  for variables substitution
```
f'normal String {variable} normal string'
```
```
name = 'Magnus'
print(f'Hello {name}')
```


#### How does variable substitution in fstring handle non string datatype 


Ans : It automatically typecasts any value into a string before substituting its value in our current string 


```
name = 'Magnus'
age = 10
print(f'Hi {name}, Your age is {age}') 
```


In [ ]:
name = 'Magnus'
age = 10
print(f'Hi {name}, Your age is {age}') 

Hi Magnus, Your age is 10


### Excersise 3 /
### Repeat excersise-2 using fstrings



In [ ]:
a = "LOL"
b = 23

print(f'My ID is {a} and I like {b}')

My ID is LOL and I like 23


In [ ]:
piece = ["Rook","Pawn","Bishop"]
points = [5,1,3]
def Chess(piece,points):
   for i in range(len(points)):
      a = piece[i]
      b = points[i]
      print(f'{a} is worth {b} Points')

In [ ]:
Chess(piece,points)

Rook is worth 5 Points
Pawn is worth 1 Points
Bishop is worth 3 Points


### Additional functionalities available in fstrings 

1. Variable printing with variable name 

For Eg : in the previous code, if we have to print what is the value of a and b aat every iteration, then we might need to write a code similar to below 

```
for i in range(len(points)):
      a = piece[i]
      b = points[i]
      print(f'At Iteration i = {i}, a = {a} , b = {b}')
```

In [ ]:
piece = ["Rook","Pawn","Bishop"]
points = [5,1,3]
for i in range(len(points)):
      a = piece[i]
      b = points[i]
      print(f'At Iteration i = {i}, a = {a} , b = {b}')

At Iteration i = 0, a = Rook , b = 5
At Iteration i = 1, a = Pawn , b = 1
At Iteration i = 2, a = Bishop , b = 3


Instread of the above, we can just use `=` operator within fstring variable substitution so that python can automatically print those variables with their variable name 

#### Syntax :
`f'{varName=}'`

Eg:

if we need to print `a=10`  in the code below, then we can just use 
```
a = 10
print(f'{a=}')
```

Output
```
a=10
```

### String Formatting 
---

printing the st

In [ ]:
10/3


3.3333333333333335

### How can I print a floating point number with fixed precision 
This can be done using a command called as `format`

For Eg :


```
a = 10/3  ## full precision
a_2decimal = "{:.2f}".format(a)
```

In [ ]:
a = 10/3  ## full precision
a_2decimal = "{:.2f}".format(a)
print(a_2decimal)




3.33


The way in which the formater works is described as 
`"{:k.pf}".format(3.3333333333333)`

- k :  How much space should I reserve for this string 
- pf : The decimal point precision \

In [ ]:
print("{:7.3f}".format(a))

  3.333


## Repeat Strings 
--- 

Similar to addition of Strings ( Concatensation ) , we can also perform repetiion of charecters using the `*` operator with the string 

```
a = "*"
b = a*5
print(b)
```
will result in 
```
*****
```

### Excersie - 4
---

So you have Two lists 
- list 1 is names of Students 
- List 2 is Marks of Students

```
names = ["Alex","Magnus","Studentwithlongname1" ,"Studentwithlonngname2"]
Marks = [96.321456997874,99.999999999,87.7564756,56.978954563453]
```


Use , forloops string formatting to generate the output below
```
S.No   |  Names                  | Marks   |
--------------------------------------------
0      |  Alex                   |   96.32 |
1      |  Magnus                 |  100.00 |
2      |  Studentwithlongname1   |   87.76 |
3      |  Studentwithlonngname2  |   56.98 |
--------------------------------------------
       |  Average                |   85.26 |
       |  Max                    |  100.00 |
       |  Min                    |   56.98 |
```





In [ ]:
#@title Solution - Problem 4
### A function to print a Matrix of marks

def MatrixOfMarks(names, marks):
  ## Here we need to compute maximum length in the name String 
  b=[]
  for i in names:
    b.append(len(i))
  MaximumLength = max(b)
  def printline():
    print("-" * (7+1+MaximumLength+4+11))

  print("{:7}".format("S.No")+"|"+"{:{len}}".format("  Names",len=MaximumLength+4)+"|"+"{:8}".format(" Marks")+" |")
  printline()
  
  for index,(nams, maks) in enumerate(zip(Names,Marks)):
    print("{:7}".format(f"{index}")+"|"+"{:{len}}".format( f"  {nams}",len=MaximumLength+4)+"|"+  "{:8.2f}".format(maks)+" |")
  
  printline()

  ## Compute Average of marks
  Average = sum(marks)/len(marks)
  Maximum = max(marks)
  Minimum = min(marks)
  print("{:7}".format("")+"|"+"{:{len}}".format("  Average",len=MaximumLength+4)+"|"+  "{:8.2f}".format(Average)+" |")
  print("{:7}".format("")+"|"+"{:{len}}".format("  Max",len=MaximumLength+4)+"|"+  "{:8.2f}".format(Maximum)+" |")
  print("{:7}".format("")+"|"+"{:{len}}".format("  Min",len=MaximumLength+4)+"|"+  "{:8.2f}".format(Minimum)+" |")


Names = ["Alex","Magnus","Studentwithlongname1" ,"Studentwithlonngname2"]
Marks = [96.321456997874,99.999999999,87.7564756,56.978954563453]
MatrixOfMarks(Names, Marks)

### Excersise 5:
---

Write a Student class , with the following details 
- Should have a constructor , which should accept `(studentName,Marks)`
   Student("Magnus",99.8757)
inside constructor, Append (add) the name to a List called as `names` within the class
Append the marks to the Marks List 
You need to have one more list which is for S.No, and you should assign serial numbers accordingly to students in which they were added


- This class should have a function called as `print_summary()` , which will print the output identical as the previous excersise.

In [ ]:
Names = ["Alex","Magnus","Studentwithlongname1" ,"Studentwithlonngname2"]
Marks = [96.321456997874,99.999999999,87.7564756,56.978954563453]

 



def printOperation2(w,h):
  for k in range(h):
    for i in range(w):
      print('-',end='')
    print('')



def printOperation3():
  for i in range(len(Marks)):
    a = Names[i]
    a += 1
    b = Marks[i]
    b += 1
    if a and b < i == "Studentwithlonngname2":
      break
    else:
      continue
    c = [1,2,3,4]
    for k in range(len(c)):
      c = c[k]
      c += 1
    if c < k == "56.978954563453":
      break
    else:
      continue
      return("{:7}".format("{c}")+"|"+"{:28}".format("  {a}")+"|"+"{:7}".format(" {b}")+"|")



def printOperation4(w,h):
  for k in range(h):
    for i in range(w):
      print('-',end='')
    print('')
class MaxMinAverage:
  mark1 = Names
  mark2 = Marks
  MARK = []
  
  def __init__(self,mark1,mark2,):
    self.mark1 = mark1
    self.mark2 = mark2
    self.MARK.append(self.mark1)
    self.MARK.append(self.mark2)
    
  def getTotalMark(self):
    a = self.mark1 + self.mark2 
    return a

  def Average(self):
    a = self.mark1 + self.mark2 
    return a/2

  def MaxMark(self):
    maxMark = 0

    for a in self.MARK:
      if a > maxMark:
        maxMark = a
    
    return maxMark
  
  def MinimumMark(self):
    minMark = 999999999
  
    for b in self.MARK:
      if b < minMark:
          minMark = b

    return minMark

  